test/reference/reference_test.go

```diff
@@ -0,0 +1,104 @@
+package reference
+
+import (
+	"os"
+	"path"
+	"testing"
+
+	"gen/github.com/woven-planet/go-zserio/testdata/reference_modules/core/instantiations"
+	"gen/github.com/woven-planet/go-zserio/testdata/reference_modules/core/types"
+	"gen/github.com/woven-planet/go-zserio/testdata/reference_modules/testobject1/testobject"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	zserio "github.com/woven-planet/go-zserio"
+)
+
+func testWorkspace(filePath string) string {
+	return path.Join(os.Getenv("TEST_SRCDIR"), os.Getenv("TEST_WORKSPACE"), filePath)
+}
+
+var (
+	// ReferenceFilePath is the path to the input data.
+	ReferenceFilePath string = testWorkspace(os.Getenv("TESTDATA_BIN"))
+)
+
+func TestRoundtripFromZserio(t *testing.T) {
+	// Given
+	want, err := os.ReadFile(ReferenceFilePath)
+	require.NoError(t, err)
+
+	// When
+	var testObject testobject.TestObject
+	require.NoError(t, zserio.Unmarshal(want, &testObject), "unmarshal")
+	got, err := zserio.Marshal(&testObject)
+	require.NoError(t, err, "marshal")
+
+	// Then
+	assert.Equal(t, want, got)
+}
+
+func TestEqualValues(t *testing.T) {
+	// Given
+	bytes, err := os.ReadFile(ReferenceFilePath)
+	require.NoError(t, err)
+
+	// When
+	var got testobject.TestObject
+	require.NoError(t, zserio.Unmarshal(bytes, &got))
+
+	// Then
+	assert.Equal(t, want(), got)
+}
+
+func want() testobject.TestObject {
+	var d testobject.TestObject
+	d.Parameter1 = 7
+	d.Struct1 = instantiations.InstantiatedTemplateStruct{
+		Parameter: d.Parameter1,
+		Field: types.ValueWrapper{
+			Parameter:   d.Parameter1,
+			Value:       72,
+			OtherValue:  121,
+			EnumValue:   types.ColorBLUE,
+			Description: "test data",
+		},
+	}
+	d.Color1 = types.ColorRED
+
+	d.Parameter2 = 12
+	for i := 0; i < 22; i++ {
+		d.StructArray = append(d.StructArray, instantiations.InstantiatedTemplateStruct{
+			Parameter: d.Parameter2,
+			Field: types.ValueWrapper{
+				Parameter:   d.Parameter2,
+				Value:       int32(i + 100),
+				OtherValue:  int8(i),
+				Description: "some dummy description",
+				// FIXME @aignas 2022-02-15: is this a bug? How can we reserialize this back to the correct binary but not have this field set?
+				// EnumValue:   types.ColorBLACK,
```

Comment:
This should be investigated a little bit more.

Reply:
See the zserio file definition:
`Color enumValue if parameter == 7;`
`Color` is only available if the parameter `d.Parameter2` is set to 7. Since this is not the case, the field is ignored, and not written.